src/tasks/python/PythonExtensionHelper.ts

```diff
@@ -0,0 +1,116 @@
+        return new Promise((resolve, reject) => {
+            vscode.tasks.onDidEndTask(async e => {
+                if (e.execution.task === prelaunchTask) {
+                    // There is no way to know the result of the completed task, so a best guess is to check if the container is running.
+                    const containerRunning = await cliDockerClient.inspectObject(containerName, { format: "{{.State.Running}}" });
+
+                    if (containerRunning === "false") {
+                        reject("Failed to attach the debugger, please see the terminal output for more details.");
+                    }
+
+                    const maxRetriesCount = 20;
+                    let retries = 0;
+                    let created = false;
+
+                    // Look for the magic string below in the log file with a retry every 0.5 second for a maximum of 10 seconds.
+                    // TODO: Should be gone as soon as the retry logic is part of the Python debugger/extension.
+                    while (++retries < maxRetriesCount && !created) {
+                        if (fse.existsSync(debuggerSemaphorePath)) {
+                            const contents = fse.readFileSync(debuggerSemaphorePath);
+
+                            created = contents.toString().indexOf("Starting server daemon on") >= 0;
+                            if (created) {
+                                break;
+                            }
+                        }
+
+                        await delay(500);
+                    }
+
+                    if (created) {
+                        resolve();
+                    } else {
+                        reject("Failed to attach the debugger within the alotted timeout.");
+                    }
+                }});
+        })
+    }
+
+    export function getRemotePtvsdCommand(target: PythonFileTarget | PythonModuleTarget, args?: string[], options?: DebugLaunchOptions): string {
+        let fullTarget: string;
+
+        if ((target as PythonFileTarget).file) {
+            fullTarget = (target as PythonFileTarget).file;
+        } else if ((target as PythonModuleTarget).module) {
+            fullTarget = `-m ${(target as PythonModuleTarget).module}`;
+        } else {
+            throw new Error("One of either module or file must be provided.");
+        }
+
+        options = options || {};
+        options.host = options.host || "0.0.0.0";
+        options.port = options.port || PythonDefaultDebugPort;
+        options.wait = !!options.wait;
+        args = args || [];
+
+        return `/pydbg/ptvsd --host ${options.host} --port ${options.port} ${options.wait ? "--wait" : ""} ${fullTarget} ${args.join(" ")}`;
+    }
+
+    export function getLauncherFolderPath(): string {
+        const pyExt = vscode.extensions.getExtension("ms-python.python");
+
+        if (!pyExt) {
+            throw new Error("For debugging Python apps in a container to work, the Python extension must be installed.");
```

Comment:
I suggest either stating the name and/or the id (`ms-python.python` in this error message? When I search for "python" in the [VS Code Marketplace](https://marketplace.visualstudio.com/vscode), there are quite a few hits out there.

Reply:
Thanks for the feedback. I agree the message should include more guidance.

Reply:
Alternatively, ensure that we display an error notification with an action that takes the user to the extension in the marketplace (or the extension in the extensions tab, which I believe there's an API for).

Reply:
I changed that. to Phil's suggestion, by showing an error message with an action that takes the user to the extension page in VSCode.